### ISSUE TITLE
Can't get tactrace.pl working

### ISSUE THREAD

**Issue by CraigEC** [MarcJHuber/event-driven-servers#14]
I’m trying to port an old installation/config to the new tacacs_plus-ng and have been running into problems I THINK with the ruleset stuff.
I tried to run the tactrace.pl to better understand the problem and have been side bared into a a different issue.
I cannot get tactrace.pl to run

This is as far as I’ve gotten and am now rather lost.

[root@authtorls2 perl]# ./tactrace.pl --help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at ./tactrace.pl line 12.
Compilation failed in require at ./tactrace.pl line 12.
BEGIN failed--compilation aborted at ./tactrace.pl line 12.

For reference, libmavis.so DOES exist.

[root@authtorls2 perl]# ls -al /usr/local/lib64/
total 228
drwxr-xr-x.  5 root root    106 Oct 17 16:38 .
drwxr-xr-x. 12 root root    131 Oct 13 12:49 ..
drwxr-xr-x.  2 root root      6 Jan 10  2022 bpf
lrwxrwxrwx   1 root root     13 Oct 13 13:10 libmavis.so -> libmavis.so.0
lrwxrwxrwx   1 root root     15 Oct 13 13:10 libmavis.so.0 -> libmavis.so.0.1
-rwxr-xr-x   1 root root 226568 Oct 13 13:10 libmavis.so.0.1
drwxr-xr-x   2 root root   4096 Oct 13 13:10 mavis
drwxr-xr-x   3 root root     18 Oct 17 16:38 perl5
[root@authtorls2 perl]#

At this point I have no clue how to further debug tactrace.pl
I’m going to return to the docs and try to figure out what might be happening with the ruleset without any tactrace.pl hints.

Worst case I’ll have to abandon tacacs_plus-ng and return to tasacs_plus.

Thanks,
-craig

**Comment by MarcJHuber** [comment]
Hi,

could you please share more details on your environment, OS version, OS distribution et al.? I'd like to run this in a VM for further evaluation, if possible. Line 12 is actually the custom Scm.pm module.

I did actually expect that tac_plus.pl could raise issues, that's why this isn't part of the default built process. 

Thanks,

Marc

**Comment by CraigEC** [comment]
This is a Redhat Linux version 9.0  (In actual fact it’s the Oracle free version of the same – guaranteed to be matching)

It is a VM system who should hopefully only be running tac_plus-ng.

I’m not really a PERL guy.  I had to build the module under spawnd/perl  with
Perl Makefile.pl
Make
Make test
Make install
To get around the initial cryptic failure.

It looks like the desired files are around, but I’m not certain how perl seeks what it requires.


On the upside, I think I authentication working.. some lucky guesses and creative config guesses..

Would be nice to get the trace working for future issues.

Thanks,
-craig
---------------------------------------------------------
Craig Campbell  ***@***.***
Waterfront International Limited
95 Wellington St. West
Suite 1900
Toronto, Ontario
Canada
M5J 2N7
416 368-6500

**Comment by CraigEC** [comment]
I did a fresh pull.

The version I got is,

***@***.*** perl]# /usr/local/sbin/tac_plus-ng -v
tac_plus-ng version a1b922b69908e66e3e48027c40a0b4597b064315/PCRE2/DES
***@***.*** perl]#

In …/event-driven-servers/
./configure --with-pcre2
make clean
make
make install


To test try running the command

/usr/local/sbin/tac_plus-ng -f /usr/local/etc/tac_plus-ng_ldap.cfg

To test LDAP functionality

/usr/local/bin/mavistest -d -1 /usr/local/etc/tac_plus-ng_ldap.cfg tac_plus-ng TACPLUS SomeUserName SomeUserPassword

All worked properly.  (i.e. I didn’t break anything)

Then went to
/event-driven-servers/tac_plus-ng/perl

Here are the results..

***@***.*** perl]# ls
GNUmakefile  tactrace.pl

So let’s see if it ‘just works…

***@***.*** perl]# ./tactrace.pl --help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at ./tactrace.pl line 12.
Compilation failed in require at ./tactrace.pl line 12.
BEGIN failed--compilation aborted at ./tactrace.pl line 12.

Nope..
Maybe I need -help?

***@***.*** perl]# ./tactrace.pl -help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at ./tactrace.pl line 12.
Compilation failed in require at ./tactrace.pl line 12.
BEGIN failed--compilation aborted at ./tactrace.pl line 12.

Nope..
Maybe I need to make?


***@***.*** perl]# pwd
/root/src/Linux9/Oct192022/event-driven-servers/tac_plus-ng/perl

***@***.*** perl]# ls
GNUmakefile  tactrace.pl

***@***.*** perl]# make
cd ../../spawnd/perl && perl Makefile.PL && make && sudo make install
WARNING: NAMEi is not a known parameter.
'NAMEi' is not a known MakeMaker parameter name.
Warning: Guessing NAME [perl] from current directory name.
Generating a Unix-style Makefile
Writing Makefile for perl
Writing MYMETA.yml and MYMETA.json
make[1]: Entering directory '/root/src/Linux9/Oct192022/event-driven-servers/spawnd/perl'
cp Scm.pm blib/lib/Scm.pm
cp receiver.pl blib/lib/receiver.pl
Running Mkbootstrap for perl ()
chmod 644 "perl.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- perl.bs blib/arch/auto/perl/perl.bs 644
"/usr/bin/perl" "/usr/share/perl5/vendor_perl/ExtUtils/xsubpp"  -typemap '/usr/share/perl5/ExtUtils/typemap'  Scm.xs > Scm.xsc
mv Scm.xsc Scm.c
gcc -c  -I../.. -D_REENTRANT -D_GNU_SOURCE -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fwrapv -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -g   -DVERSION=\"0.01\" -DXS_VERSION=\"0.01\" -fPIC "-I/usr/lib64/perl5/CORE"   Scm.c
rm -f blib/arch/auto/perl/perl.so
gcc  -lpthread -shared -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -L/usr/local/lib -fstack-protector-strong  Scm.o  -o blib/arch/auto/perl/perl.so  \
   -L/usr/local/lib64 -lmavis -lperl   \

chmod 755 blib/arch/auto/perl/perl.so
make[1]: Leaving directory '/root/src/Linux9/Oct192022/event-driven-servers/spawnd/perl'
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- perl.bs blib/arch/auto/perl/perl.bs 644
Files found in blib/arch: installing files in blib/lib into architecture dependent library tree
Installing /usr/local/lib64/perl5/5.32/auto/perl/perl.so
Appending installation info to /usr/lib64/perl5/perllocal.pod
sudo install -m 0755 tactrace.pl /usr/local/bin/tactrace.pl


***@***.*** perl]# ./tactrace.pl -help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at ./tactrace.pl line 12.
Compilation failed in require at ./tactrace.pl line 12.
BEGIN failed--compilation aborted at ./tactrace.pl line 12.
***@***.*** perl]#
***@***.*** perl]#
***@***.*** perl]#
***@***.*** perl]# ./tactrace.pl -help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at ./tactrace.pl line 12.
Compilation failed in require at ./tactrace.pl line 12.
BEGIN failed--compilation aborted at ./tactrace.pl line 12.
***@***.*** perl]#

Nope
But wait, the make installed in /usr/local/bin   Maybe if I run it there?

***@***.*** perl]# /usr/local/bin/tactrace.pl --help
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at /usr/local/bin/tactrace.pl line 12.
Compilation failed in require at /usr/local/bin/tactrace.pl line 12.
BEGIN failed--compilation aborted at /usr/local/bin/tactrace.pl line 12.

That didn’t work..

***@***.*** perl]# /usr/local/bin/tactrace.pl
Can't load '/usr/local/lib64/perl5/5.32/auto/Scm/Scm.so' for module Scm: libmavis.so: cannot open shared object file: No such file or directory at /usr/lib64/perl5/DynaLoader.pm line 193.
at /usr/local/bin/tactrace.pl line 12.
Compilation failed in require at /usr/local/bin/tactrace.pl line 12.
BEGIN failed--compilation aborted at /usr/local/bin/tactrace.pl line 12.
***@***.*** perl]#

Nor that…

And yet the file it is apparently complaining about does exist..

***@***.*** perl]# ls -al /usr/local/lib64/perl5/5.32/auto/Scm/Scm.so
-r-xr-xr-x 1 root root 70784 Oct 18 07:12 /usr/local/lib64/perl5/5.32/auto/Scm/Scm.so

And there do seem to be plentiful libmavis.so around…
***@***.*** perl]# find /usr/local/ -name "libmavis*"
/usr/local/lib64/libmavis.so.0
/usr/local/lib64/libmavis.so
/usr/local/lib64/mavis/libmavis_remote.so
/usr/local/lib64/mavis/libmavis_system.so
/usr/local/lib64/mavis/libmavis_groups.so
/usr/local/lib64/mavis/libmavis_anonftp.so
/usr/local/lib64/mavis/libmavis_asciiftp.so
/usr/local/lib64/mavis/libmavis_limit.so
/usr/local/lib64/mavis/libmavis_cache.so
/usr/local/lib64/mavis/libmavis_auth.so
/usr/local/lib64/mavis/libmavis_log.so
/usr/local/lib64/mavis/libmavis_external.so
/usr/local/lib64/mavis/libmavis_userdb.so
/usr/local/lib64/mavis/libmavis_tee.so
/usr/local/lib64/mavis/libmavis_null.so
/usr/local/lib64/mavis/libmavis_tacinfo_cache.so
/usr/local/lib64/libmavis.so.0.1
***@***.*** perl]#

Not sure what bit of glue is missing – or what sort of operator error I have performed.
I’ve tried to include answers to the obvious questions.
tac_plus-ng IS running fine, so it can’t be anything major..

Hope this helps.

Thanks,
-craig





---------------------------------------------------------
Craig Campbell  ***@***.***
Waterfront International Limited
95 Wellington St. West
Suite 1900
Toronto, Ontario
Canada
M5J 2N7
416 368-6500

**Comment by MarcJHuber** [comment]
Hi Craig,

I've pushed another commit, please pull and retry. Unfortunately, I couldn't test this with Oracle Linux 9 due to VirtualBox issues ...

Thanks,

Marc

**Comment by CraigEC** [comment]
Hi Marc,
                Looks like that did it.

I did need to do an additional make in

../event-driven-servers/tac_plus-ng/perl

As the tactrace.pl wasn’t updated in /usr/local/bin

Thanks so much!

-craig
---------------------------------------------------------
Craig Campbell  ***@***.***
Waterfront International Limited
95 Wellington St. West
Suite 1900
Toronto, Ontario
Canada
M5J 2N7
416 368-6500

**Comment by MarcJHuber** [comment]
Hi Craig,

great, thanks a lot!

Cheers,

Marc

**Comment by CraigEC** [comment]
I’m not sure what your VirtualBox issues were, but I found that installing Linux 9 required more resources than previously.
Specifically I had to increase memory to 3GB and hard disk to 100GB

Perhaps that will help you in the future.

-craig

---------------------------------------------------------
Craig Campbell  ***@***.***
Waterfront International Limited
95 Wellington St. West
Suite 1900
Toronto, Ontario
Canada
M5J 2N7
416 368-6500
